### ISSUE TITLE
Unexpected ConnectionClosed in keepalive ping task

### ISSUE THREAD

**Issue by Harmon758** [aaugustin/websockets#551]
Using websockets 7.0 on Python 3.7, this can occur semi-frequently:
```
Unexpected exception in keepalive ping task
Traceback (most recent call last):
  File "C:\Program Files\Python37\lib\site-packages\websockets\protocol.py", line 989, in keepalive_ping
	ping_waiter, self.ping_timeout, loop=self.loop
  File "C:\Program Files\Python37\lib\asyncio\tasks.py", line 416, in wait_for
	return fut.result()
websockets.exceptions.ConnectionClosed: WebSocket connection is closed: code = 1006 (connection closed abnormally [internal]), no reason
```
This sometimes is followed by:
```
Error in data transfer
Traceback (most recent call last):
  File "C:\Program Files\Python37\lib\site-packages\websockets\protocol.py", line 674, in transfer_data
	message = yield from self.read_message()
  File "C:\Program Files\Python37\lib\site-packages\websockets\protocol.py", line 742, in read_message
	frame = yield from self.read_data_frame(max_size=self.max_size)
  File "C:\Program Files\Python37\lib\site-packages\websockets\protocol.py", line 846, in read_data_frame
	pong_waiter.set_result(None)
asyncio.base_futures.InvalidStateError: invalid state
```
The first error seems to contradict the documentation in `keepalive_ping`:
https://github.com/aaugustin/websockets/blob/8fc78fee48d52bb3c690e925bad0825613319296/src/websockets/protocol.py#L967 I believe this is because pending keepalive pings are now aborted with `ConnectionClosed` in 7.0 (#467).

I think the latter error is because the ping is [marked done with `set_exception`](https://docs.python.org/3/library/asyncio-future.html#asyncio.Future.set_exception) in `abort_keepalive_pings` where `ConnectionClosed` is raised:
https://github.com/aaugustin/websockets/blob/8fc78fee48d52bb3c690e925bad0825613319296/src/websockets/protocol.py#L1148 and
https://github.com/aaugustin/websockets/blob/8fc78fee48d52bb3c690e925bad0825613319296/src/websockets/protocol.py#L840 attempts to set the result to `None`, raising [an error](https://docs.python.org/3/library/asyncio-exceptions.html#asyncio.InvalidStateError) [when the Future is already done](https://docs.python.org/3/library/asyncio-future.html#asyncio.Future.set_result).

**Comment by lgrahl** [comment]
I've seen it happen too but I so far was unable to figure out why it occurs.

**Comment by AFO-UYI** [comment]
Same issue here. I was developing a twitch bot with websockets. Everything worked fine. But at the update to v7.0, the 1006 error often appears. Even on the connection. Sometimes works fine, sometimes doesnt. I guess I must downgrade.

**Comment by aaugustin** [comment]
Sorry, I haven't had time to investigate yet. Definitely a valid issue considering the multiple reports.

**Comment by WojciechMigda** [comment]
> I have this issue as well and had to downgrade to version 6.0 to work around it.

@fenhl downgrade is not necessary, since with version 7.0 passing `ping_interval=None` to the `Client` or `Server` disables ping.

**Comment by rhiller** [comment]
Very similar errors on 7.0 with python 3.7.2

```
04/08/2019 18:09:34.090 WARNING websockets.protocol None Unexpected exception in keepalive ping task
Traceback (most recent call last):
  File "/home/ronbo/deploy/p5.server/lib64/python3.7/site-packages/websockets/protocol.py", line 984, in keepalive_ping
    ping_waiter = yield from self.ping()
  File "/home/ronbo/deploy/p5.server/lib64/python3.7/site-packages/websockets/protocol.py", line 583, in ping
    yield from self.ensure_open()
  File "/home/ronbo/deploy/p5.server/lib64/python3.7/site-packages/websockets/protocol.py", line 658, in ensure_open
    ) from self.transfer_data_exc
websockets.exceptions.ConnectionClosed: WebSocket connection is closed: code = 1011 (unexpected error), no reason
```

I got a bunch of these all at the same on different connections when the ISP at the other end flaked out.

Typically in such situations I see:
```
WebSocket connection is closed: code = 1006 (connection closed abnormally [internal]), no reason
```

**Comment by bithon** [comment]
same issue with Python 3.5.2 on Linux

**Comment by bithon** [comment]
> > I have this issue as well and had to downgrade to version 6.0 to work around it.
> 
> @fenhl downgrade is not necessary, since with version 7.0 passing `ping_interval=None` to the `Client` or `Server` disables ping.
but then you will not recognize a broken connection, so you have to care about ping/pong too

**Comment by WojciechMigda** [comment]
@bithon correct. Fortunately, in my usecase the server I am connecting too offerts a `heartbeat` message in its higher level protocol. But if you expect inbound traffic with some bound frequency then not receiving any input over websockets channel might serve such purpose too, I guess.

**Comment by vazir** [comment]
Experiencing similar stuff in the PONG. Pypy latest version.

websockets.exceptions.ConnectionClosed: WebSocket connection is closed: code = 1000 (OK), reason = STOPING WS SERVICE AFTER 30000 
Error in data transfer 
Traceback (most recent call last): 
 File "/usr/local/lib/python3.6/site-packages/websockets/protocol.py", line 674, in transfer_data 
   message = yield from self.read_message() 
 File "/usr/local/lib/python3.6/site-packages/websockets/protocol.py", line 742, in read_message 
   frame = yield from self.read_data_frame(max_size=self.max_size) 
 File "/usr/local/lib/python3.6/site-packages/websockets/protocol.py", line 835, in read_data_frame 
   yield from self.pong(frame.data) 
 File "/usr/local/lib/python3.6/site-packages/websockets/protocol.py", line 618, in pong 
   yield from self.write_frame(True, OP_PONG, data) 
 File "/usr/local/lib/python3.6/site-packages/websockets/protocol.py", line 919, in write_frame 
   yield from self.ensure_open() 
 File "/usr/local/lib/python3.6/site-packages/websockets/protocol.py", line 646, in ensure_open 
   ) from self.transfer_data_exc 
websockets.exceptions.ConnectionClosed: WebSocket connection is closed: code = 1006 (connection closed abnormally [internal]), no reason 
Error in data transfer 
Traceback (most recent call last): 
 File "/usr/local/lib/python3.6/site-packages/websockets/protocol.py", line 674, in transfer_data 
   message = yield from self.read_message() 
 File "/usr/local/lib/python3.6/site-packages/websockets/protocol.py", line 742, in read_message 
   frame = yield from self.read_data_frame(max_size=self.max_size) 
 File "/usr/local/lib/python3.6/site-packages/websockets/protocol.py", line 835, in read_data_frame 
   yield from self.pong(frame.data) 
 File "/usr/local/lib/python3.6/site-packages/websockets/protocol.py", line 618, in pong 
   yield from self.write_frame(True, OP_PONG, data) 
 File "/usr/local/lib/python3.6/site-packages/websockets/protocol.py", line 919, in write_frame 
   yield from self.ensure_open() 
 File "/usr/local/lib/python3.6/site-packages/websockets/protocol.py", line 646, in ensure_open 
   ) from self.transfer_data_exc 
websockets.exceptions.ConnectionClosed: WebSocket connection is closed: code = 1006 (connection closed abnormally [internal]), no reason

**Comment by aaugustin** [comment]
Indeed, websockets doesn't handle well the situation where a ping is aborted because the connection was lost.

As noted by @Harmon758 in this bug report, this could manifest in two places:

1) When websockets aborts a ping that the keepalive task is waiting on;
2) When websockets processes a pong that matches an aborted ping.

The two commits in PR #627 address these two issues.

**Comment by aaugustin** [comment]
The issue here is noise in logs. There are no other ill effects.

**Comment by Abbyyan** [comment]
> The issue here is noise in logs. There are no other ill effects.

But the return code here is `$?=1`

**Comment by Whokmw** [comment]
Is there a solution? I am having the same issue and still can't see a way out.

**Comment by aaugustin** [comment]
This issue was resolved. If you're still seeing it with a recent version of websockets, then:

- either it's a new issue
- or it reoccurred

Please file a new ticket and provide the complete traceback.

**Comment by Whokmw** [comment]
Everything works, it's all good now.

**Comment by primelionzard** [comment]
@Whokmw I have the same questions as @somduttasinha , what did you end up having to do?
